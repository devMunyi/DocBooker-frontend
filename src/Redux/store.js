import { configureStore, applyMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import doctorReducer from './reducers/doctors';
import reservationReducer from './reducers/reservations';
import selectedDoctorReducer from './reducers/selectedDoctor';
import allDoctors from './reducers/allDoctors';
import user from './reducers/user';


export default configureStore(
  {
    reducer: {
      doctor: doctorReducer,
      reservations: reservationReducer,
      selectedDoctor: selectedDoctorReducer,
      doctors: allDoctors,
      user,
    },
  },
  applyMiddleware(logger),
);
