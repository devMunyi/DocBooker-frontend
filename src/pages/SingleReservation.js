import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useParams, useNavigate } from 'react-router-dom';
import { DeleteOutlined, EditOutlined, SyncOutlined } from '@ant-design/icons';
import axios from 'axios';
import { toast } from 'react-toastify';
import NavBar from '../components/NavBar';
import { fetchReservation } from '../Redux/actions/reservations';
import { fetchSingleDoctor } from '../Redux/actions/doctors';

function SingleReservation() {
  const [isProcessing, setIsProcessing] = useState(false);
  const dispatch = useDispatch();
  const { reservationId, doctorId } = useParams();
  const navigate = useNavigate();

  const userId = localStorage.getItem('user')
    ? JSON.parse(localStorage.getItem('user')).id
    : null;

  if (!userId === null) {
    navigate('/login');
  }

  useEffect(() => {
    console.log('Called');
    dispatch(fetchReservation({ userId, doctorId, reservationId }));
    dispatch(fetchSingleDoctor({}));
  }, [dispatch, userId, doctorId, reservationId]);

  const { reservation, isFetchingData } = useSelector(
    (state) => state.reservations,
  );

  localStorage.setItem('reservation', JSON.stringify(reservation));

  const handleUpdateBtnClick = () => {
    navigate(`/reservation/edit/${reservation.id}`);
  };

  const handleDelete = async () => {
    try {
      setIsProcessing(true);

      const response = await axios.delete(
        `http://localhost:3000/api/users/${userId}/doctors/${doctorId}/reservations/${reservationId}`,
        {
          headers: {
            'Content-Type': 'application/json',
          },
        },
      );
      const { data } = response;
      if (response.status === 204) {
        toast.warning('Reservation deleted successfully!');

        navigate(`/reservations/${userId}/doctor/${doctorId}`);
      } else {
        throw new Error(data.message || 'Failed to delete reservation');
      }
    } catch (error) {
      toast.error(error.message || 'Failed to delete reservation');
    } finally {
      setIsProcessing(false);
    }
  };

  return (
    <div className="page OneReservationPage">
      <NavBar />
      <main className="content">
        <h1>Reservation Details</h1>
        {isFetchingData ? (
          <div
            className="d-flex justify-content-center align-items-center"
            style={{
              width: '100%',
            }}
          >
            <SyncOutlined spin style={{ fontSize: '32px', color: '#589d1a' }} />
          </div>
        ) : (
          <div className="row">
            <div className="col-md-9">
              <table className="table table-bordered">
                <tbody className="">
                  <tr>
                    <th scope="row">Doctor Name</th>
                    <td>{reservation.doctorname}</td>
                  </tr>
                  <tr>
                    <th scope="row">Specialization</th>
                    <td>{reservation.specialization}</td>
                  </tr>
                  <tr>
                    <th scope="row">Date</th>
                    <td>{reservation.date}</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div className="col-md-3">
              <div className="d-flex flex-column gap-3 actions justify-content-center align-items-center">
                <button
                  onClick={handleUpdateBtnClick}
                  type="button"
                  className="btn btn-outline-info btn-sm d-flex justify-content-center align-items-center gap-2"
                >
                  {' '}
                  <EditOutlined />
                  {' '}
                  <span>Update</span>
                </button>
                {isProcessing ? (
                  <button
                    className="btn btn-outline-light"
                    type="button"
                    disabled
                  >
                    <span
                      className="spinner-border spinner-border-sm"
                      role="status"
                      aria-hidden="true"
                    />
                    <span className="visually-hidden">Loading...</span>
                  </button>
                ) : (
                  <button
                    onClick={handleDelete}
                    type="button"
                    className="btn btn-outline-danger btn-sm d-flex justify-content-center align-items-center gap-2"
                  >
                    <DeleteOutlined />
                    {' '}
                    <span>Delete</span>
                  </button>
                )}
              </div>
            </div>
          </div>
        )}
      </main>
    </div>
  );
}

export default SingleReservation;
