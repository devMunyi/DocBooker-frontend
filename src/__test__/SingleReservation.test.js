import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import store from '../Redux/store';
import SingleReservation from '../pages/SingleReservation';

it('SingleReservation component renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <SingleReservation />
        </Router>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
