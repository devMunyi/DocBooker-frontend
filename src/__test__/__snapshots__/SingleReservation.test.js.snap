// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleReservation component renders correctly 1`] = `
<div
  className="page OneReservationPage"
>
  <div
    className="menu-toggle "
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="hamburger"
    >
      <span />
    </div>
  </div>
  <aside
    className="sidebar "
  >
    <div
      className="dropdown d-flex justify-content-center align-items-center border-bottom"
    >
      <button
        aria-expanded="false"
        className="btn btn-secondary dropdown-toggle border-0"
        data-bs-toggle="dropdown"
        style={
          Object {
            "backgroundColor": "#589d1a",
          }
        }
        type="button"
      >
        <span
          aria-label="user"
          className="anticon anticon-user border rounded-circle p-1"
          role="img"
          style={
            Object {
              "fontSize": "32px",
            }
          }
        >
          <svg
            aria-hidden="true"
            data-icon="user"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
            />
          </svg>
        </span>
        <br />
        <small
          className="text-light"
        >
          Welcome Sam
        </small>
      </button>
      <ul
        className="dropdown-menu"
      >
        <li>
          <span
            className="dropdown-item"
            onClick={[Function]}
            role="presentation"
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Logout
          </span>
        </li>
      </ul>
    </div>
    <nav
      className="menu mt-4"
    >
      <a
        aria-current="page"
        className="menu-item active"
        href="/"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "underline",
          }
        }
      >
        Doctors
      </a>
      <a
        className="menu-item"
        href="/reserve/1/doctor/1"
        onClick={[Function]}
      >
        My Reservations
      </a>
      <a
        className="menu-item"
        href="/reservations/2/doctor/4"
        onClick={[Function]}
      >
        Add Reservation
      </a>
      <a
        className="menu-item"
        href="/add/doctor"
        onClick={[Function]}
      >
        Add Doctor
      </a>
      <a
        className="menu-item"
        href="/add-doctor"
        onClick={[Function]}
      >
        Delete Doctor
      </a>
    </nav>
  </aside>
  <main
    className="content"
  >
    <h1>
      Reservation Details
    </h1>
    <div
      className="row"
    >
      <div
        className="col-md-9"
      >
        <table
          className="table table-bordered"
        >
          <tbody
            className=""
          >
            <tr>
              <th
                scope="row"
              >
                Doctor Name
              </th>
              <td />
            </tr>
            <tr>
              <th
                scope="row"
              >
                Specialization
              </th>
              <td />
            </tr>
            <tr>
              <th
                scope="row"
              >
                Date
              </th>
              <td />
            </tr>
          </tbody>
        </table>
      </div>
      <div
        className="col-md-3"
      >
        <div
          className="d-flex flex-column gap-3 actions justify-content-center align-items-center"
        >
          <button
            className="btn btn-outline-info btn-sm d-flex justify-content-center align-items-center gap-2"
            onClick={[Function]}
            type="button"
          >
             
            <span
              aria-label="edit"
              className="anticon anticon-edit"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="edit"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                />
              </svg>
            </span>
             
            <span>
              Update
            </span>
          </button>
          <button
            className="btn btn-outline-danger btn-sm d-flex justify-content-center align-items-center gap-2"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-label="delete"
              className="anticon anticon-delete"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="delete"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                />
              </svg>
            </span>
             
            <span>
              Delete
            </span>
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;
